Add repel mod to the osu ruleset

ggliv opened the pull request:
Initially brought up in #18294.

This mod effectively does the opposite of the existing magnetised mod, pushing hitobjects away from your cursor instead of pulling them toward it. ~~The gradual hitobject repositioning shared between the two mods is implemented in a new abstract class, `OsuEaseHitObjectPositionsMod` (I am not 100% sure on the naming convention I should use here, please let me know if it should be changed).~~ (No longer abstracted as of [2fe34f1](https://github.com/ppy/osu/pull/18607/commits/2fe34f188f4c6d4294cd98f2f4d18a921f5949a1))

`OsuHitObjectGenerationUtils.calculatePossibleMovementBounds` is exposed to allow the mod to ensure sliders do not move outside of the bounds of the playfield, ~~and a new parameter (defaulted to true) is added to determine whether the returned `RectangleF` should account for the radius of the slider.~~ (Reverted as of [6e883a6](https://github.com/ppy/osu/pull/18607/commits/6e883a69d9e98947053646d6f166afe0bf4a277a)) The slider clamping logic is reused from [`clampSliderToPlayfield`](https://github.com/ppy/osu/blob/79a17fc86a9bc68065a7aad4943f7c2fcd00ecb8/osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils_Reposition.cs#L203-L211).

https://user-images.githubusercontent.com/47010459/172435116-6b09a48c-6fd6-4abb-9e5f-e96ab3565184.mp4

ggliv committed:
initial implementation

ggliv committed:
add test, fix formatting, expose easing function

ggliv committed:
clamp sliders, expose slider bounds function

ggliv committed:
update tests to match new bindable names

ggliv committed:
add newline

mk56-spn commented:
Ive played it a bit and it feels very akward, which i guess its fair enough but the maximum repel  strength needs to be cut down to whats currently .75 strength, anything above that is just not really reasonable

apollo-dw commented:
I'd personally expect repel to be a checkbox / switch in Magnetised, rather than a whole separate mod

frenzibyte commented:
Having it a separate mod is consistent with Grow/Deflate, WindUp/WindDown, etc. I think that's fine for now.

ggliv committed:
scale down repulsion strength

bdach commented on osu.Game.Rulesets.Osu/Utils/OsuHitObjectGenerationUtils_Reposition.cs:
```diff
@@ -212,10 +212,13 @@ private static void applyDecreasingShift(IList<OsuHitObject> hitObjects, Vector2
         /// If the slider is larger than the playfield, the returned <see cref="RectangleF"/> may have negative width/height.
         /// </remarks>
-        private static RectangleF calculatePossibleMovementBounds(Slider slider)
+        public static RectangleF CalculatePossibleMovementBounds(Slider slider, bool accountForFollowCircleRadius = true)
```
What is the param for? Would prefer it to not exist if it can be avoided.

Also the name is wrong, I think? It's not accounting for the "follow circle". It's accounting for the *breadth of the actual slider*. `slider.Path` is a zero-width path, the bounds of the actual object are not only influenced by the control points of the path, but also by the width (or radius) of the slider path itself.

ggliv replied:
The param was added so that sliders and hitcircles could share the behavior of repelling to the point that their centers touched the playfield border. When considering the slider radius, a slider will repel until its radius touches the playfield border, not its center. As far as I can tell, this is a non-issue with the scaled-down easing strength, so I'm ok with reverting. This may be changed if #18559 gets merged anyways.

You are correct on the naming point, thank you for your explanation.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs:
```diff
@@ -0,0 +1,58 @@
+    {
+        public override string Name => "Repel";
+        public override string Acronym => "RP";
+        public override IconUsage? Icon => FontAwesome.Solid.ExpandArrowsAlt;
```
Please don't specify mod icons. We've decided to avoid doing so for the time being. Icons will be designed for mods specifically and properly at a later date, rather than picking the least worst option from fontawesome selection.

ggliv committed:
revert slider radius parameter addition

ggliv committed:
don't specify icon

ggliv committed:
replace easement with easing

ggliv committed:
shamelessly copy osumodmagnetised

peppy commented:
Is it intended that the repulsion gets strong the *further away objects are to the cursor*? This to be seems counter-intuitive, and hasn't been mentioned anywhere, so I wanted to clarify before going any further.

peppy committed:
Improve description

ggliv commented:
I didn't explicitly intend for that to be the case while putting this together, but I think it works well enough gameplay-feel-wise. Using a constant strength makes repulsion while the cursor is close to a hitobject feel jerky, in my opinion, though I could just be used to the original feel at this point.

<details>
<summary>Here's the diff for constant strength if you want to give it a try.</summary>

```diff
diff --git a/osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs b/osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs
index 052c7128f6..371e2edfd0 100644
--- a/osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs
+++ b/osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs
@@ -51,24 +51,20 @@ public void Update(Playfield playfield)
 
             foreach (var drawable in playfield.HitObjectContainer.AliveObjects)
             {
-                float x = Math.Clamp(2 * drawable.X - cursorPos.X, 0, OsuPlayfield.BASE_SIZE.X);
-                float y = Math.Clamp(2 * drawable.Y - cursorPos.Y, 0, OsuPlayfield.BASE_SIZE.Y);
+                var destination = drawable.Position - 100 * Vector2.NormalizeFast(cursorPos - drawable.Position);
+                destination = Vector2.Clamp(destination, Vector2.Zero, OsuPlayfield.BASE_SIZE);
 
                 if (drawable.HitObject is Slider thisSlider)
                 {
                     var possibleMovementBounds = OsuHitObjectGenerationUtils.CalculatePossibleMovementBounds(thisSlider);
 
-                    x = possibleMovementBounds.Width < 0
-                        ? x
-                        : Math.Clamp(x, possibleMovementBounds.Left, possibleMovementBounds.Right);
-
-                    y = possibleMovementBounds.Height < 0
-                        ? y
-                        : Math.Clamp(y, possibleMovementBounds.Top, possibleMovementBounds.Bottom);
+                    destination = Vector2.Clamp(
+                        destination,
+                        new Vector2(possibleMovementBounds.Left, possibleMovementBounds.Top),
+                        new Vector2(possibleMovementBounds.Right, possibleMovementBounds.Bottom)
+                    );
                 }
 
-                var destination = new Vector2(x, y);
-
                 switch (drawable)
                 {
                     case DrawableHitCircle circle:

```

</details>

peppy commented:
To make it clear, I was implying it should (if going by the physically correct definition) be *opposite* of what it is right now, not constant.

ggliv committed:
operate on vectors instead of vector components

ggliv committed:
change default strength back to 0.5

ggliv committed:
modify damp length to effectively invert repulsion strength

ggliv committed:
Merge branch 'master' into osu-mod-repel

ggliv committed:
add nullable directive

mk56-spn commented:
I'll give it a play

mk56-spn commented:
sorry for the late response, feels really nice now, its smooth and it doesn't sort of getting annoying to hit when you get close to the circle. predictable movement, well done. gameplay wise i feel it's good for a merge now

peppy committed:
Merge branch 'master' into osu-mod-repel

ggliv committed:
update new usage of CalculatePossibleMovementBounds

ggliv committed:
use cursor position instead of destination for dampLength calculation

the destination vector is clamped within playfield borders, we want dampLength to be based on distance from the cursor.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs:
```diff
@@ -0,0 +1,97 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Would be preferable to have all new code have NRT enabled. The problem here is going to be `gameplayClock`, but I think the way to go here is to annotate it as nullable and then add a `Debug.Assert(gameplayClock != null)` in `easeTo()`.

bdach commented on osu.Game.Rulesets.Osu/Mods/OsuModRepel.cs:
```diff
@@ -0,0 +1,97 @@
+        public override ModType Type => ModType.Fun;
+        public override string Description => "Hit objects run away!";
+        public override double ScoreMultiplier => 1;
+        public override Type[] IncompatibleMods => new[] { typeof(OsuModAutopilot), typeof(OsuModWiggle), typeof(OsuModTransform), typeof(ModAutoplay), typeof(OsuModMagnetised) };
```
This mod is declaring incompatibility with 5 others, but 4 of those 5 are not declaring incompatibility with `OsuModRepel` themselves in the other direction (`OsuModMagnetised` being the exception).

peppy replied:
You still missed autoplay..

peppy replied:
and relax.

ggliv replied:
Yikes.

Relax was intentionally excluded though, #17319 isn't an issue when hitobjects are being pushed away.

ggliv committed:
reciprocate mod incompatibility

ggliv committed:
enable NRT again

peppy committed:
Add missed incompatilibity rules

peppy committed:
Merge branch 'master' into osu-mod-repel
